File: Practica1/C++/quicksort.cpp
#include <bits/stdc++.h>
#include "../GeneradorAleatorio/generator.h"
using namespace std;
void crear(int *&A,int n)
{
	A=new int[n];
}
void Aleatorio(int& n,int* arreglo){
    srand(time(nullptr));
    int num;
    for (int i =0;i<n;i++){
        num = 1+rand()%n;
        arreglo[i] = num;
    }
    cout<<endl;

}

void print(int& n, int* arreglo){
    for(int i=0;i<n;i++){
        cout<<arreglo[i]<<" ";
    }
    cout<<endl;
}

int dividir(int arr[], int inicio, int fin)
{
    int pivot = arr[fin];
    int i = (inicio - 1);

    for (int j = inicio; j <= fin - 1; j++) {

        if (arr[j] <= pivot) {

            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[fin]);
    return (i + 1);
}

int divrandom(int arr[], int low, int high)
{
    srand(time(NULL));
    int random = low + rand() % (high - low);

    swap(arr[random], arr[high]);

    return dividir(arr, low, high);
}

void quickSort(int arr[], int inicio, int fin)
{
    if (inicio < fin) {


        int pi = divrandom(arr, inicio, fin);

        quickSort(arr, inicio, pi - 1);
        quickSort(arr, pi + 1, fin);
    }
}

int main(int argc, char const *argv[])
{
    int n=10000;
	double t0;
	double time1;
	int k=0;
	ofstream fs("Tiempos/time_Quick.txt");
	srand(time(NULL));
	for(int i=n;i<=100000;i=i+10000){
		int *A;
		crear(A,i);
		read(A,i,k);
		t0=clock();
		quickSort(A,0,i-1);
		t0=clock()-t0;
		time1= ((double)t0)/CLOCKS_PER_SEC;
		k++;
		fs<<time1<<endl;
	}
    return 0;
}

File: Practica1/Java/quicksort.java
import java.util.*;
public class quicksort {
    int dividir(int arr[],int inicio,int fin){
        int pivot = arr[fin];
        int x= inicio-1;
        for(int i=inicio;i<fin;i++ ){
            if(arr[i]< pivot){
                x++;
                int swap = arr[x];
                arr[x] = arr[i];
                arr[i] = swap; 
            }
        }
        int swap = arr[x+1];
        arr[x+1] = arr[fin];
        arr[fin] = swap;
        return x+1;
    }
    int divrandom(int arr[],int inicio,int fin){
        Random rand = new Random();
        int pivot = rand.nextInt(fin-inicio)+inicio;
        int swap = arr[pivot];
        arr[pivot] = arr[fin];
        arr[fin] = swap;
        return dividir(arr,inicio,fin);
    }
    void qsort(int arr[],int inicio,int fin){
        if(inicio<fin){
            int part = divrandom(arr,inicio,fin);
            qsort(arr,inicio,part-1);
            qsort(arr,part+1,fin);
        }
    }

    static void print(int arr[]) 
    { 
        int n = arr.length; 
        for (int i=0; i<n; ++i) 
            System.out.print(arr[i]+" "); 
        System.out.println(); 
    } 

    static void generar(int arr[],int n){
        int rand = 0;
        for(int i=0;i<n;i++){
            rand=(int)(Math.random()*n+1);
            arr[i] = rand;
        }

    }

    public static void main(String[] args) {
        int n = 0;
        long t1, t2, tt;
        Scanner entrada = new Scanner(System.in);
        n = entrada.nextInt();
        int[] arr = new int[n];
        generar(arr,n);
        //System.out.println("Arreglo Desordenado");
        quicksort quick = new quickal(); 
        t1=System.currentTimeMillis();
        quick.qsort(arr,0,n-1);
        t2= System.currentTimeMillis();
        tt=t2-t1;
        System.out.println("Tiempo de ejecucion: "+tt);
        //print(arr); 
    }
}
